### ISSUE TITLE
Duplicate tax issue

### ISSUE THREAD

**Issue by ekomurz** [mollie/Magento#130]
After updating the Mollie module to 5.4.0 the checkout process was suddenly changed:
* in our site, products are tax included. On checkout, the tax was added once more and also displayed extra
* without any documentation, the Mollie fee was added to all order confirmations ?

Did we overlook something ?
For now, rollback to 5.3.0 is working okay.

**Comment by Frank-Magmodules** [comment]
Hi @ekomurz , thank you for the message. 

First Point; Mollie doesn't change anything in the Checkout/Tax logics but please [contact](https://www.magmodules.nl/contact-us.html) us directly if you are sure this is caused by the Mollie extension.

Second Point; I'm gonna try to reproduce this and let you know.

Frank

**Comment by Ig0r-M** [comment]
Same issue happening.
Due to above Mollie is directly interfering with other payment methods.

**Comment by Ig0r-M** [comment]
On checkout `Tax: ` section appearing regardless of option to inc./excl. tax is cause by the total `mollie_mpm_payment_fee_tax`. Will provide more details if i find where and how exactly this is done.

**Comment by Frank-Magmodules** [comment]
Hi @Ig0r-M , Thanks, we are investigating the issue at this moment as well.

**Comment by Frank-Magmodules** [comment]
Hi @Ig0r-M and @ekomurz , thanks again for detailed issue report. 
The issue has been resolved in the [v5.5.0](https://github.com/mollie/Magento/releases/tag/v5.5.0) release. 
We are closing this issue now but please feel free to reopen the issue if this still occurs.

**Comment by PascalBrouwers** [comment]
I am on the latest version. Commit https://github.com/mollie/Magento/commit/2e990ca765208c41bd936cd09cbff5c574409f61
Why is this still not fixed?

**Comment by Frank-Magmodules** [comment]
Hi @PascalBrouwes, this issue has been fixed since v5.5.0. 

Are you experiencing the same issue; that extra tax is added over the total value? Are you updating from a previous Mollie version, if yes, from what version? Also, can you share your Magento1 version? 

Thank you!

**Comment by PascalBrouwers** [comment]
I am running the latest version: 5.5.6. Together with Magento 1.9.3.10.
After enabling the Mollie module products are added with additional tax. E.g. adding a product including tax of 399.95 becomes 469.36.
Disabling all totals of the module seems to prevent it: `mollie_mpm_payment_fee` and `mollie_mpm_payment_fee_tax`

**Comment by peterjaap** [comment]
Coincidentally, I'm also running into issues with `mollie_mpm_payment_fee` and `mollie_mpm_payment_fee_tax`. When these are enabled, my "Free shipping" rule isn't honored.

I have a rule where free shipping should be applied > 100 euro. This is with these totals enabled (incorrect):

![image](https://user-images.githubusercontent.com/431360/80200211-97ae1600-8622-11ea-982e-5ea59fcc9bf0.png)

This is it with those totals commented out in `config.xml` (correct):

![image](https://user-images.githubusercontent.com/431360/80200299-b8766b80-8622-11ea-849d-8cd34c7d28ab.png)

**Comment by seansan** [comment]
Re open issue?

**Comment by PascalBrouwers** [comment]
Mollie sees it's in issue in my case with another payment provider module (MSP) and are looking into it.

**Comment by PascalBrouwers** [comment]
Seems to be a Core Magento bug. This fixes it: https://stackoverflow.com/a/11356906/6389747
Unfortunately nothing can be done about this in the Mollie module.

**Comment by Frank-Magmodules** [comment]
Hi @PascalBrouwers ,

Yes exactly, that's the URL Michiel shared with you this morning. However, we are happy to look for a suitable solution for other Merchants. It looks like that adding multiple custom total is the root cause of this issue. We are looking for a way to identify this so we can include this in the self-test and documentation.

Keep you posted!

**Comment by seansan** [comment]
Tried applying this patch from stack .. this caused other issues (like empty totals or no subtotals in cart)

https://github.com/mollie/Magento/issues/130#issuecomment-618919678

**Comment by PascalBrouwers** [comment]
Something like this:
```
class YourClass extends Mage_Sales_Model_Quote_Address_Total_Collector
{
protected function _getSortedCollectorCodes() {
    if (Mage::app()->useCache('config')) {
        $cachedData = Mage::app()->loadCache($this->_collectorsCacheKey);
        if ($cachedData) {
            return unserialize($cachedData);
        }
    }
    $configArray = $this->_modelsConfig;
    // invoke simple sorting if the first element contains the "sort_order" key
    reset($configArray);
    $element = current($configArray);
    if (isset($element['sort_order'])) {
        uasort($configArray, array($this, '_compareSortOrder'));
        $sortedCollectors = array_keys($configArray);
    } else {
        foreach ($configArray as $code => $data) {
            foreach ($data['before'] as $beforeCode) {
                if (!isset($configArray[$beforeCode])) {
                    continue;
                }
                $configArray[$code]['before'] = array_merge(
                        $configArray[$code]['before'],
                        $configArray[$beforeCode]['before']);
                $configArray[$code]['before'] = array_unique(
                        $configArray[$code]['before']);
                $configArray[$beforeCode]['after'] = array_merge(
                        $configArray[$beforeCode]['after'], array($code),
                        $data['after']);
                $configArray[$beforeCode]['after'] = array_unique(
                        $configArray[$beforeCode]['after']);
            }
            foreach ($data['after'] as $afterCode) {
                if (!isset($configArray[$afterCode])) {
                    continue;
                }
                $configArray[$code]['after'] = array_merge(
                        $configArray[$code]['after'],
                        $configArray[$afterCode]['after']);
                $configArray[$code]['after'] = array_unique(
                        $configArray[$code]['after']);
                $configArray[$afterCode]['before'] = array_merge(
                        $configArray[$afterCode]['before'], array($code),
                        $data['before']);
                $configArray[$afterCode]['before'] = array_unique(
                        $configArray[$afterCode]['before']);
            }
        }
        //uasort($configArray, array($this, '_compareTotals'));
        $res = "";
        foreach ($configArray as $code => $data) {
            foreach ($data['before'] as $beforeCode) {
                if (!isset($configArray[$beforeCode])) {
                    continue;
                }
                $res = $res . "$code $beforeCode\n";
            }
            foreach ($data['after'] as $afterCode) {
                if (!isset($configArray[$afterCode])) {
                    continue;
                }
                $res = $res . "$afterCode $code\n";
            }
        }
        file_put_contents(Mage::getBaseDir('tmp')."/graph.txt", $res);
        $sortedCollectors=explode("\n",shell_exec('tsort '.Mage::getBaseDir('tmp')."/graph.txt"),-1);           
    }
    if (Mage::app()->useCache('config')) {
        Mage::app()
                ->saveCache(serialize($sortedCollectors),
                        $this->_collectorsCacheKey,
                        array(Mage_Core_Model_Config::CACHE_TAG));
    }
    return $sortedCollectors;
}
}
```

**Comment by seansan** [comment]
Cool! We had a 1-on-1 chat ... with Pascal. This exact code above seems to fix our issue. Tested on production with both Mollie and Adyen. @peterjaap can maybe confirm

![image](https://user-images.githubusercontent.com/652395/80600517-cc5c0c00-8a2c-11ea-945a-cd5106f7f2d8.png)

**Comment by peterjaap** [comment]
Shelling out just to run a topological sort seems a bit .... unsafe. Not to mention that it won't work on all servers (especially shared servers).

I'd rather swap out the `tsort` shell command with a proper PHP topological sorting implementation like https://github.com/vaimo/topological-sort.

**Comment by michielgerritsen** [comment]
@PascalBrouwers @seansan @peterjaap,

I've opened a pull request which adds extra before & afters to the `tax_subtotal` total. This makes sure that the totals stay in de correct place. Could you try this out and give feedback to this? I already tried this on the test environment of @PascalBrouwers as I had access to this.

**Comment by seansan** [comment]
@michielgerritsen I can check this ... 

As we use 2 PSPs in parallel ... would this change then solve it for both or only Mollie you think?

**Comment by michielgerritsen** [comment]
@seansan This bug is caused by the fact that 2 modules (2 PSPs in your case) try to add a custom total. Magento does not handle this case well and messes up the sort order of the totals. This chance tries to let the orders intact. So yes this solution should fix it for both payment methods.

**Comment by seansan** [comment]
- Reverted previous fix to original file
- Tested ideal on mollie
- Tested ideal on ayden
- Tested paypal express on paypal

All seem working ;) So test looks succesfull until now, will monitor backend orders

**Comment by peterjaap** [comment]
Good work everybody!!

**Comment by Marvin-Magmodules** [comment]
Fix has been merged into master and releases as [v5.5.7](https://github.com/mollie/Magento/releases/tag/v5.5.7). Thanks all for the feedback, testing and sharing developer environments! Much appreciated. We are closing this issue now but please feel free to reopen the issue if this still occurs.
